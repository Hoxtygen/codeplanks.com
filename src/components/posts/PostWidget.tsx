import Link from "next/link";
import React, { useState, useEffect } from "react";
import { getRecentPosts, getSimilarPosts } from "../../services";
import { BlogError, PostWidgetProps, PWidget } from "../../typedefs";
import { formatDate } from "../../utils";

export default function PostWidget({ categories, slug }: PostWidgetProps) {
  const [relatedPosts, setRelatedPosts] = useState<PWidget[]>([]);
  const [error, setError] = useState<BlogError | null>(null);
  useEffect(() => {
    if (slug) {
      getSimilarPosts(categories!, slug)
        .then((result) => setRelatedPosts(result))
        .catch((err: BlogError) => setError(err));
    } else {
      getRecentPosts()
        .then((result) => setRelatedPosts(result))
        .catch((err: BlogError) => setError(err));
    }
  }, [slug]);

  return (
    <div className="bg-white shadow-lg rounded-lg p-8 mb-8">
      <h3 className="text-2xl mb-4 font-semibold border-b pb-4">
        {slug ? "Related Posts" : "Recent Posts"}{" "}
      </h3>
      {error && <p>{error.error}</p>}
      {relatedPosts.map((post) => (
        <div key={post.title} className="flex items-center w-full">
          <div className="w-16 flex-none">
            <img
              src={post.featuredImage.url}
              alt=""
              height="60px"
              width="60px"
              className="align-middle rounded-full"
            />
          </div>
          <div className="flex-grow ml-4">
            <p className="text-gray-500 font-xs">
              {formatDate(post.createdAt)}
            </p>
            <Link
              key={post.title}
              href={`/post/${post.slug}`}
              className="text-md"
            >
              {post.title}
            </Link>
          </div>
        </div>
      ))}
    </div>
  );
}
