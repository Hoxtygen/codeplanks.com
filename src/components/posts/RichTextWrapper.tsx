import React, { Fragment, useEffect } from "react";
import { RichText } from "@graphcms/rich-text-react-renderer";
import Link from "next/link";
import Image from "next/image";
import { RichContextProps } from "../../typedefs";
import Prism from "prismjs";
import "prismjs/plugins/line-numbers/prism-line-numbers";
import "prismjs/themes/prism-tomorrow.css";
import "prismjs/plugins/line-numbers/prism-line-numbers.css";

export default function RichTextWrapper({ content }: RichContextProps) {
  useEffect(() => {
    Prism.highlightAll();
  }, []);
  return (
    <div>
      <RichText
        content={content.raw}
        renderers={{
          a: ({ children, openInNewTab, href, ...rest }) => {
            if (href?.match(/^https?:\/\/|^\/\//i)) {
              return (
                <a
                  href={href}
                  target={openInNewTab ? "_blank" : "_self"}
                  rel="noopener noreferrer"
                  {...rest}
                  className="underline decoration-solid decoration-sky-500 hover:decoration-blue-400"
                >
                  {children}
                </a>
              );
            }
            return (
              <Link href={href as string}>
                <a data-testid="blog-link" {...rest}>
                  {children}{" "}
                </a>
              </Link>
            );
          },
          h1: ({ children }) => <h1 className="text-white">{children}</h1>,
          h2: ({ children }) => (
            <h2 className="text-xl font-semibold mb-4">{children}</h2>
          ),
          h3: ({ children }) => <h3>{children}</h3>,
          h4: ({ children }) => <h4>{children}</h4>,
          h5: ({ children }) => <h5>{children}</h5>,
          h6: ({ children }) => <h6>{children}</h6>,
          italic: ({ children }) => <i>{children} </i>,
          p: ({ children }) => (
            <p data-testid="paragraph" className="mb-8">
              {children}
            </p>
          ),
          bold: ({ children }) => <strong>{children}</strong>,
          ul: ({ children }) => (
            <ul aria-label="unordered list" className="list-disc list-inside">
              {children}
            </ul>
          ),
          ol: ({ children }) => (
            <ol
              aria-label="ordered list"
              data-testid="ordered list"
              className="list-decimal list-inside"
            >
              {children}
            </ol>
          ),
          li: ({ children }) => <li aria-label="list item"> {children}</li>,
          underline: ({ children }) => <u>{children} </u>,
          code_block: ({ children }) => {
            return (
              <pre
                className="line-numbers language-none"
                data-testid="code-block"
              >
                <code className="language-javascript">{children}</code>
              </pre>
            );
          },
          code: ({ children }) => (
            <code className="language-javascript">{children}</code>
          ),
          table: ({ children }) => (
            <table data-testid="table">{children} </table>
          ),
          table_head: ({ children }) => <thead>{children} </thead>,
          table_body: ({ children }) => <tbody> {children}</tbody>,
          table_row: ({ children }) => <tr> {children}</tr>,
          list_item_child: ({ children }) => <Fragment> {children}</Fragment>,
          blockquote: ({ children }) => (
            <figure data-testid="quote">
              <blockquote>
                <p className="text-lg font-medium">{children}</p>
              </blockquote>
            </figure>
          ),
          img: ({ src, altText, height, width }) => (
            <div
              style={{
                width: "100%",
                display: "flex",
                justifyContent: "center",
              }}
            >
              <Image
                src={src as string}
                alt={altText}
                height={height}
                width={width}
                objectFit="cover"
              />
            </div>
          ),
        }}
      />
    </div>
  );
}
