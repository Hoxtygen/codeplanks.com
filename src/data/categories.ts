export const testCategories = {
  categories: [
    {
      id: "cl8yjhbie2dj00bn1qq2wml1d",
      name: "Web Development",
      slug: "web-dev",
    },
    { id: "cl8zulkgd3hoi0bn75ccbcybd", name: "React", slug: "react" },
    {
      id: "cl8zuqc5u3k290an220qa4y4c",
      name: "Flutter",
      slug: "flutter",
    },
    { id: "cl8zuqsae3k7x0an203v29ztg", name: "NodeJS", slug: "node-js" },
  ],
};
