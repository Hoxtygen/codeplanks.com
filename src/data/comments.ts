export const testComments = {
  comments: [
    {
      id: "cl9ptdgr712um0alp2qqae0rg",
      name: "Machiavelli",
      email: "nacuabjkj dfd",
      comment:
        "Lorem ipsum dolor sit amet consectetur adipisicing elit. Quam ex commodi voluptatibus laboriosam est? Soluta, odio natus vel blanditiis mollitia impedit expedita temporibus sit cum dolores enim eveniet eaque dolorum repellat labore hic magnam non assumenda laboriosam.\n",
      createdAt: "2022-10-30T16:25:30.706593+00:00",
    },
    {
      id: "cl9ptkjjl15ob0an5upjhw2mo",
      name: "Julian Ward",
      email: "julina",
      comment:
        "Lorem ipsum dolor sit amet consectetur adipisicing elit. Quam ex commodi voluptatibus laboriosam est? Soluta, odio natus vel blanditiis mollitia impedit expedita temporibus sit cum dolores enim eveniet eaque dolorum repellat labore hic magnam non assumenda laboriosam.\n",
      createdAt: "2022-10-30T16:25:30.706593+00:00",
    },
    {
      id: "cl9v9w8zje3820an5hdkao5mu",
      name: "Marijuana Trump",
      email: "marijuana.trump@sterling.ng",
      comment:
        "Lorem ipsum dolor sit amet consectetur, adipisicing elit. Harum voluptates odit iste, id est aliquid molestias consequuntur nemo obcaecati aspernatur non alias doloribus laborum eveniet tempora excepturi quia in voluptatum dolor corporis adipisci odio! Architecto recusandae dolorum ullam nihil obcaecati.\n",
      createdAt: "2022-10-30T16:25:30.706593+00:00",
    },
    {
      id: "cl9wfird6s03n0blsonpo2sa6",
      name: "Job-Search",
      email: "b@example.com",
      comment: "Lorem makes sense at all times",
      createdAt: "2022-10-30T16:25:30.706593+00:00",
    },
  ],
};
