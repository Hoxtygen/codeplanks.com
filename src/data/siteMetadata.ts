const siteMetadata = {
  companyName: "Codeplanks",
  phoneNumber: "(+234) 806-018-4972",
  siteUrl: "https://codeplanks.com",
  siteLogo: "https://example.com/images/logo.png",
  siteLogoSquare: "https://example.com/images/logox200.png",
  email: "adenwas@gmail.com",
  twitter: "https://twitter.com/hoxtygen",
  twitterHandle: "@hoxtygen",
  facebook: "https://www.facebook.com/example.page",
  linkedin: "https://www.linkedin.com/in/wasiu-idowu/",
  ogImageUrl: "https://www.placholderimageurl.com",
};

export default siteMetadata;
