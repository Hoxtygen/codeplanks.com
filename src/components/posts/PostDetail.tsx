/* eslint-disable @next/next/no-img-element */
import React from "react";

import RichTextWrapper from "./RichTextWrapper";
import { Post } from "../../typedefs";
import { formatDate } from "../../utils";

export default function PostDetail({ post }: Post) {
  return (
    <div className="bg-white shadow-lg rounded-lg lg:p-8 pb-12 mb-8">
      <div className="relative overflow-hidden shadow-md mb-6">
        <img
          src={post.featuredImage.url}
          alt={post.title}
          className="object-top h-full w-full rounded-t-lg"
        />
      </div>
      <div className="px-4 lg:px-0">
        <div className="flex items-center mb-8 w-full">
          <div className="flex items-center r mb-4 lg:mb-0 w-full lg:w-auto mr-8">
            <img
              src={post.author.photo.url}
              alt={post.author.name}
              className="align-middle rounded-full"
              height="50px"
              width="50px"
              data-testid="author-image"
            />
            <p
              data-testid="author-name"
              className="inline align-middle text-gray-700 ml-2 text-lg"
              aria-label="Author name"
            >
              {post.author.name}
            </p>
          </div>
          <div className="font-medium text-gray-700">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="h-6 w-6 inline mr-2 text-pink-500"
              fill="none"
              viewBox="0 0 24 24"
              stroke="currentColor"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
              />
            </svg>
            <span data-testid="published-date" aria-label="Publication date">
              {formatDate(post.createdAt)}
            </span>
          </div>
        </div>
        <h2
          data-testid="post-title"
          className="mb-8 text-3xl font-semibold"
          data-name={post.title}
        >
          {post.title}
        </h2>
        <RichTextWrapper content={post.content} />
      </div>
    </div>
  );
}
