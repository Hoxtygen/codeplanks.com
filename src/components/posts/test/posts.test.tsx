import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import Home from "../../../pages";
import { testPosts } from "../../../data/posts";
import PostCard from "../PostCard";

describe("Blog posts", () => {
  describe("Home", () => {
    it("renders the Home component", () => {
      const { container } = render(<Home posts={testPosts.posts} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe("PostCard component", () => {
    const post = testPosts.posts[0];
    it("should render PostCard component", () => {
      const { container } = render(<PostCard post={post} />);
      expect(container).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    it("should render a post", () => {
      render(<PostCard post={post} />);
      const excerpt = screen.getByTestId("excerpt");
      const heading = screen.getByRole("heading");
      const images = screen.getAllByRole("img");
      const link = screen.getByRole("link");
      const postDate = screen.getByTestId("postDate");
      const postTitle = screen.getByText(post.title);
      const readMore = screen.getByTestId("readmore");
      const readMoreText = screen.getByText(/Continue reading/i);

      expect(excerpt).toBeInTheDocument();
      expect(excerpt).toMatchSnapshot();
      expect(heading).toBeInTheDocument();
      expect(heading).toMatchSnapshot();
      expect(images).toHaveLength(2);
      expect(images).toMatchSnapshot();
      expect(link).toBeInTheDocument();
      expect(link).toMatchSnapshot();
      expect(postDate).toBeInTheDocument();
      expect(postDate).toMatchSnapshot();
      expect(postTitle).toBeInTheDocument();
      expect(postTitle).toMatchSnapshot();
      expect(postTitle.textContent).toBe("Promises");
      expect(readMore).toBeInTheDocument();
      expect(readMore).toMatchSnapshot();
      expect(readMoreText).toBeInTheDocument();
      expect(readMoreText).toMatchSnapshot();
    });
  });
});
