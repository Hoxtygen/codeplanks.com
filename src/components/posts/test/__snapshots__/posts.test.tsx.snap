// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog posts Home renders the Home component 1`] = `
<div>
  <div
    class="container mx-auto px-10 mb-8"
  >
    <div
      class="grid grid-cols-1 lg:grid-cols-12 gap-12"
    >
      <div
        class="lg:col-span-8 col-span-1"
      >
        <div
          class="bg-white shadow-lg rounded-lg p-0 lg:p-8 pb-12 mb-8"
        >
          <div
            class="relative overflow-hidden shadow-md pb-80 mb-6"
          >
            <img
              alt="Promises"
              class="object-top absolute h-80 w-full object-cover shadow-lg rounded-t-lg lg:rounded-lg"
              src="https://media.graphassets.com/zfNO9rUaTSWv46vUnqUw"
            />
          </div>
          <h1
            class="transition duration-700 text-center mb-8 cursor-pointer hover:text-pink-600 text-3xl font-semibold"
          >
            <a
              href="/post/javascript-promises"
            >
              Promises
            </a>
          </h1>
          <div
            class="block lg:flex text-center items-center justify-center mb-8 w-full"
          >
            <div
              class="flex items-center justify-center mb-4 lg:mb-0 w-full lg:w-auto mr-8"
            >
              <img
                alt="Wasiu Idowu"
                class="align-middle rounded-full"
                height="50px"
                src="https://media.graphassets.com/x1acQHzbTsivb1K1vieY"
                width="50px"
              />
              <p
                class="inline align-middle text-gray-700 ml-2 text-lg"
              >
                Wasiu Idowu
              </p>
            </div>
            <div
              class="font-medium text-gray-700"
            >
              <svg
                class="h-6 w-6 inline mr-2 text-pink-500"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                data-testid="postDate"
              >
                Oct 7, 2022
              </span>
            </div>
          </div>
          <p
            class="text-center text-lg text-gray-700 font-normal px-4 lg:px-20 mb-8"
            data-testid="excerpt"
          >
            JavaScript promise is used for asynchronous programming, a successor to callback.
          </p>
          <div
            class="text-center"
          >
            <span
              class="transition duration-500 transform hover:-translate-y-1 inline-block bg-black text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
              data-testid="readmore"
            >
              Continue reading
            </span>
          </div>
        </div>
        <div
          class="bg-white shadow-lg rounded-lg p-0 lg:p-8 pb-12 mb-8"
        >
          <div
            class="relative overflow-hidden shadow-md pb-80 mb-6"
          >
            <img
              alt="Promises-Part 2"
              class="object-top absolute h-80 w-full object-cover shadow-lg rounded-t-lg lg:rounded-lg"
              src="https://media.graphassets.com/O2tkABqROSLPOn89Syic"
            />
          </div>
          <h1
            class="transition duration-700 text-center mb-8 cursor-pointer hover:text-pink-600 text-3xl font-semibold"
          >
            <a
              href="/post/javascript-promises-part-2"
            >
              Promises-Part 2
            </a>
          </h1>
          <div
            class="block lg:flex text-center items-center justify-center mb-8 w-full"
          >
            <div
              class="flex items-center justify-center mb-4 lg:mb-0 w-full lg:w-auto mr-8"
            >
              <img
                alt="Wasiu Idowu"
                class="align-middle rounded-full"
                height="50px"
                src="https://media.graphassets.com/x1acQHzbTsivb1K1vieY"
                width="50px"
              />
              <p
                class="inline align-middle text-gray-700 ml-2 text-lg"
              >
                Wasiu Idowu
              </p>
            </div>
            <div
              class="font-medium text-gray-700"
            >
              <svg
                class="h-6 w-6 inline mr-2 text-pink-500"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                data-testid="postDate"
              >
                Oct 7, 2022
              </span>
            </div>
          </div>
          <p
            class="text-center text-lg text-gray-700 font-normal px-4 lg:px-20 mb-8"
            data-testid="excerpt"
          >
            JavaScript promise is used for asynchronous programming, a successor to callback.
          </p>
          <div
            class="text-center"
          >
            <span
              class="transition duration-500 transform hover:-translate-y-1 inline-block bg-black text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
              data-testid="readmore"
            >
              Continue reading
            </span>
          </div>
        </div>
      </div>
      <div
        class="lg:col-span-4 col-span-1"
      >
        <div
          class="lg:sticky relative top-8"
        >
          <div
            class="bg-white shadow-lg rounded-lg p-8 mb-8"
          >
            <h3
              class="text-2xl mb-4 font-semibold border-b pb-4"
            >
              Recent Posts
               
            </h3>
          </div>
          <div
            class="bg-white shadow-lg rounded-lg p-8 mb-8 pb-12"
          >
            <h3
              class="text-2xl mb-4 font-semibold border-b pb-4"
            >
              Categories
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Blog posts PostCard component should render PostCard component 1`] = `
<div>
  <div
    class="bg-white shadow-lg rounded-lg p-0 lg:p-8 pb-12 mb-8"
  >
    <div
      class="relative overflow-hidden shadow-md pb-80 mb-6"
    >
      <img
        alt="Promises"
        class="object-top absolute h-80 w-full object-cover shadow-lg rounded-t-lg lg:rounded-lg"
        src="https://media.graphassets.com/zfNO9rUaTSWv46vUnqUw"
      />
    </div>
    <h1
      class="transition duration-700 text-center mb-8 cursor-pointer hover:text-pink-600 text-3xl font-semibold"
    >
      <a
        href="/post/javascript-promises"
      >
        Promises
      </a>
    </h1>
    <div
      class="block lg:flex text-center items-center justify-center mb-8 w-full"
    >
      <div
        class="flex items-center justify-center mb-4 lg:mb-0 w-full lg:w-auto mr-8"
      >
        <img
          alt="Wasiu Idowu"
          class="align-middle rounded-full"
          height="50px"
          src="https://media.graphassets.com/x1acQHzbTsivb1K1vieY"
          width="50px"
        />
        <p
          class="inline align-middle text-gray-700 ml-2 text-lg"
        >
          Wasiu Idowu
        </p>
      </div>
      <div
        class="font-medium text-gray-700"
      >
        <svg
          class="h-6 w-6 inline mr-2 text-pink-500"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          data-testid="postDate"
        >
          Oct 7, 2022
        </span>
      </div>
    </div>
    <p
      class="text-center text-lg text-gray-700 font-normal px-4 lg:px-20 mb-8"
      data-testid="excerpt"
    >
      JavaScript promise is used for asynchronous programming, a successor to callback.
    </p>
    <div
      class="text-center"
    >
      <span
        class="transition duration-500 transform hover:-translate-y-1 inline-block bg-black text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
        data-testid="readmore"
      >
        Continue reading
      </span>
    </div>
  </div>
</div>
`;

exports[`Blog posts PostCard component should render a post 1`] = `
<p
  class="text-center text-lg text-gray-700 font-normal px-4 lg:px-20 mb-8"
  data-testid="excerpt"
>
  JavaScript promise is used for asynchronous programming, a successor to callback.
</p>
`;

exports[`Blog posts PostCard component should render a post 2`] = `
<h1
  class="transition duration-700 text-center mb-8 cursor-pointer hover:text-pink-600 text-3xl font-semibold"
>
  <a
    href="/post/javascript-promises"
  >
    Promises
  </a>
</h1>
`;

exports[`Blog posts PostCard component should render a post 3`] = `
[
  <img
    alt="Promises"
    class="object-top absolute h-80 w-full object-cover shadow-lg rounded-t-lg lg:rounded-lg"
    src="https://media.graphassets.com/zfNO9rUaTSWv46vUnqUw"
  />,
  <img
    alt="Wasiu Idowu"
    class="align-middle rounded-full"
    height="50px"
    src="https://media.graphassets.com/x1acQHzbTsivb1K1vieY"
    width="50px"
  />,
]
`;

exports[`Blog posts PostCard component should render a post 4`] = `
<a
  href="/post/javascript-promises"
>
  Promises
</a>
`;

exports[`Blog posts PostCard component should render a post 5`] = `
<span
  data-testid="postDate"
>
  Oct 7, 2022
</span>
`;

exports[`Blog posts PostCard component should render a post 6`] = `
<a
  href="/post/javascript-promises"
>
  Promises
</a>
`;

exports[`Blog posts PostCard component should render a post 7`] = `
<span
  class="transition duration-500 transform hover:-translate-y-1 inline-block bg-black text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
  data-testid="readmore"
>
  Continue reading
</span>
`;

exports[`Blog posts PostCard component should render a post 8`] = `
<span
  class="transition duration-500 transform hover:-translate-y-1 inline-block bg-black text-lg font-medium rounded-full text-white px-8 py-3 cursor-pointer"
  data-testid="readmore"
>
  Continue reading
</span>
`;
